linera-service/src/chain_leader.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use futures::{lock::Mutex, StreamExt};
+use linera_core::{
+    client::{ChainClient, ValidatorNodeProvider},
+    tracker::NotificationTracker,
+    worker::Reason,
+};
+use linera_storage::Store;
+use linera_views::views::ViewError;
+use std::{ops::DerefMut, sync::Arc};
+use thiserror::Error as ThisError;
+use tracing::{debug, error, warn};
+
+/// A `ChainLeader` is a process that listens to notifications from validators and reacts appropriately.
+pub struct ChainLeader<P, S> {
+    client: Arc<Mutex<ChainClient<P, S>>>,
+}
+
+#[derive(Debug, ThisError)]
+enum ChainLeaderError {
+    // #[error(transparent)]
+    // BcsHexError(#[from] BcsHexParseError),
+    // #[error("could not decode query string")]
+    // QueryStringError(#[from] hex::FromHexError),
+    // #[error(transparent)]
+    // BcsError(#[from] bcs::Error),
+    // #[error(transparent)]
+    // JsonError(#[from] serde_json::Error),
+    #[error(transparent)]
+    Internal(#[from] anyhow::Error),
+    // #[error("missing graphql operation")]
+    // MissingOperation,
+    // #[error("unsupported query type: subscription")]
+    // UnsupportedQueryType,
+    // #[error("graphql operations of different types submitted")]
+    // HeterogeneousOperations,
+    // #[error("failed to parse graphql query: {error}")]
+    // GraphQLServerError { error: String },
+    // #[error("malformed application response")]
+    // MalformedApplicationResponse,
```

Comment:
Should we have a TODO or ticket for this?

Reply:
Sorry I forgot to clean up